src/api/apiErrors.js

```diff
@@ -0,0 +1,62 @@
+/* @flow strict-local */
+import type { ApiResponseErrorData } from './transportTypes';
+import * as logging from '../utils/logging';
+
+/**
+ * Runtime class of custom API error types.
+ */
+export class ApiError extends Error {
+  data: $ReadOnly<ApiResponseErrorData>;
+  httpStatus: number;
+
+  constructor(httpStatus: number, data: $ReadOnly<ApiResponseErrorData>) {
+    super(data.msg);
+    this.data = data;
+    this.httpStatus = httpStatus;
+  }
+}
+
+/**
+ * Given a server response (allegedly) denoting an error, produce an Error to be
+ * thrown.
+ *
+ * If the `data` argument is actually of the form expected from the server, the
+ * returned error will be an {@link ApiError}; otherwise it will be a generic
+ * Error.
+ */
+export const makeErrorFromApi = (httpStatus: number, data: ?mixed): Error => {
+  // Validate `data`, and construct the resultant error object.
+  if (data !== null && typeof data === 'object') {
+    if (typeof data.msg === 'string' && data.result === 'error') {
+      if ('code' in data && typeof data.code === 'string') {
+        /* $FlowFixMe, in the unlikely event we can ever narrow types with if
+        statements on properties. */
+        return new ApiError(httpStatus, data);
+      }
+    }
+  }
+
+  // Server has responded, but the response is not a valid error-object.
+  // (This should never happen, even on old versions of the Zulip server.)
```

Comment:
I think not having `code` is actually normal pre-1.7.0.

Logic here seems fine, just adjust comment and perhaps the warning message. The logged warning will in turn help us gauge if we can forget pre-1.7.0 servers yet!

Reply:
> The existing logic _is_ incorrect, since it'll sometimes throw away a `msg` provided by a pre-1.7.0 server!

Hmm, fair enough!